### ISSUE TITLE
terraform article

### ISSUE THREAD

**Issue by davidobrien1985** [davidobrien1985/davidobrien1985.github.io#1]

**Comment by davidobrien1985** [comment]
this is a test

**Comment by davidobrien1985** [comment]
another test

**Comment by mrballcb** [comment]
I do like the ternary command syntax `(TEST ? value_if_true : value_if_false)`.  I find it to be very readable, once I equated "? ... :" with that in my head.  It seems to be present in most languages, and I just assumed that I could/should use it because I have seen lots of people using it.  Your irritation with it made me momentarily reassess that.  I thought about it and my conclusion is as long as code is readable by me later, and by any who follow me, if it's appropriate for the language I'll keep doing it.

**Comment by ddarwent** [comment]
It's worth remembering that Terraform is a single part of the Hashicorp ecosystem, so you're only seeing part of the picture by using Terraform alone. If you are using just the open source version then you need other tools for, things like secret management configuration management etc. I've used Source Control and Azure DevOps successfully to provide these features, which has transformed how I use Terraform and how my templates are constructed.

I disagree about Terraform being too simple, you can build pretty complex templates with the built-in functions (https://www.terraform.io/docs/configuration/interpolation.html), I agree count is a sticky plaster, but I believe this will be fixed in V0.12. I'd have loved to see some more examples on what ARM can do that Terraform can't (it would help to increase my ARM knowledge).

Most of my frustrations with using Terraform have come from the Azure provider, rather than Terraform, and certain parts of that frustration is due to the Azure ARM API. The lack of support for certain resources (though the same can be said for ARM e.g. Storage CORS settings) and the way certain resources have been represented e.g. the access policy for a Key Vault was part of the Key Vault resource, it's now a separate resource, which is much more flexible.

My initial impression of Terraform, over a year ago, was that it's really simple to get started (over ARM), but the Azure provider was lacking support of enough resources. There's been significant progress with the provider over the last 12 months and Terraform v0.12 is promising a lot.

Maybe it's time for my own blog post. =)

**Comment by 4c74356b41** [comment]
nah, terraform is just terrible. its limitations are endless compared to arm templates. creating weird resources which do not exist in azure (like nic to load balancer connection). plans are just horrible. its just meh over all.
ps. HCL is a definition of garbage.
